src/realm/object-store/sync/sync_session.cpp

```diff
@@ -500,21 +502,20 @@ void SyncSession::handle_error(SyncError error)
     else if (error_code.category() == realm::sync::protocol_error_category()) {
         SimplifiedProtocolError simplified =
             get_simplified_error(static_cast<sync::ProtocolError>(error_code.value()));
-        if (error.server_requests_client_reset) {
+        if (error.server_requests_action) {
             // we have specific instructions from the server concerning client resets
-            switch (*error.server_requests_client_reset) {
-                case SyncError::ClientResetModeAllowed::DoNotClientReset:
+            switch (*error.server_requests_action) {
+                case sync::ProtocolErrorInfo::Action::ClientReset:
+                    [[fallthrough]];
+                case sync::ProtocolErrorInfo::Action::ClientResetNoRecovery:
+                    simplified = SimplifiedProtocolError::ClientResetRequested;
+                    break;
+                default:
                     if (simplified == SimplifiedProtocolError::ClientResetRequested) {
                         // Server says not to do a client reset, but the client wants
                         // to do one; make a fatal (non-reset) error instead.
                         simplified = SimplifiedProtocolError::UnexpectedInternalIssue;
                     }
-                    break;
-                case SyncError::ClientResetModeAllowed::RecoveryPermitted:
-                    [[fallthrough]];
-                case SyncError::ClientResetModeAllowed::RecoveryNotPermitted:
-                    simplified = SimplifiedProtocolError::ClientResetRequested;
-                    break;
             }
         }
         switch (simplified) {
```

Comment:
do we need to update this to Do What The Server Says™️ ?

Reply:
I am not really sure what the client should do for the other actions other than client reset (`ProtocolViolation`, `ApplicationBug`, `Warning`, and `Transient`). I guess in case of `ApplicationBug` we could stop sync (maybe also for `ProtocolViolation`; the server states the user should upgrade the SDK in these cases), and retry for the rest? Any opinions?